Gems/Terrain/Code/Source/TerrainRaycast/TerrainRaycastContext.cpp

```diff
@@ -0,0 +1,369 @@
+        const AZ::Vector3 refinedMin(aabbMin.GetX(), aabbMin.GetY(), refinedMinZ);
+        const AZ::Vector3 refinedMax(aabbMax.GetX(), aabbMax.GetY(), refinedMaxZ);
+        const AZ::Aabb refinedAABB = AZ::Aabb::CreateFromMinMax(refinedMin, refinedMax);
+
+        // Check for a hit against the refined AABB.
+        float intersectionT;
+        float intersectionEndT;
+        const int intersectionResult = AZ::Intersect::IntersectRayAABB2(rayStart,
+                                                                        rayDirectionReciprocal,
+                                                                        refinedAABB,
+                                                                        intersectionT,
+                                                                        intersectionEndT);
+        if (intersectionResult == AZ::Intersect::ISECT_RAY_AABB_NONE)
+        {
+            return;
+        }
+
+        // Finally, triangulate the four terrain points and check for a hit,
+        // splitting using the top-left -> bottom-right diagonal so to match
+        // the current behavior of the terrain physics and rendering systems.
+        AzFramework::RenderGeometry::RayResult bottomLeftIntersectionResult;
+        FindNearestIntersection(rayStart,
+                                rayDirection,
+                                point0,
+                                point1,
+                                point3,
+                                bottomLeftIntersectionResult);
+
+        AzFramework::RenderGeometry::RayResult topRightIntersectionResult;
+        FindNearestIntersection(rayStart,
+                                rayDirection,
+                                point2,
+                                point3,
+                                point1,
+                                topRightIntersectionResult);
+
+        if (bottomLeftIntersectionResult)
+        {
+            result = !topRightIntersectionResult || bottomLeftIntersectionResult.m_distance < topRightIntersectionResult.m_distance ?
+                     bottomLeftIntersectionResult :
+                     topRightIntersectionResult;
+        }
+        else if (topRightIntersectionResult)
+        {
+            result = topRightIntersectionResult;
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////
+    // Recursive function that sub-divides an AABB encompasing terrain points, stopping when the size of
+    // the smallest sub-divided AABBs matches the given grid resolution, then obtains the terrain height
+    // values at each corner in order to triangulate them and then find the nearest intersection (if any)
+    // between the resulting triangles and the given ray. The AABB is only sub-divided about the x/y axes
+    // and we don't calculate any height values until we get down to an AABB that matches the resolution.
+    //
```

Comment:
Wait...  I don't think this algorithm will work correctly.  This looks like it's making the assumption that the terrain heights at the 4 corners of a grid square represent the AABB for that square?  Imagine a square that contains corners at 0, but a middle point at 100.  This would construct a 0-height AABB for that square wouldn't it, and miss the intersection with the middle point?

Reply:
We don't calculate any height values until we reach a sub-divided AABB that is the same size (in the x and y axes) as the grid dimensions; up until that point we continue using the same z values of the top level AABB that was initially passed into this recursive function, which for the current only use case is the world bounds of the terrain system (which was confirmed is guaranteed to always enclose all valid terrain points?).

So say we have a terrain world with dimensions {128, 128, 10} and terrain grid resolution 2x2, the first sub-division will yield four AABBs with dimensions {64, 64, 10}, then {32, 32, 10}, then {16, 16, 10}, etc. until we get to {2, 2, 10} - only at this point to we calculate the height values at the four corners of the smallest sub-divided AABB, after which we do one last AABB ray check against a smaller AABB of dimensions {2, 2, maxHeight - minHeight}, before finally triangulating the four points and ray casting against those.

Reply:
Another way to think of it is to imagine the terrain world bounds as a grid of columns, where the x/y extents of each column is equal to the terrain grid resolution, and the z extent (height) of each column is equal to the high of the terrain world bounds.

The first part of the algorithm is to figure out which of the columns a ray hits, but by recursively sub-diving the initial terrain world bounds in such a way that we don't have to test against every individual column, but rather progressively narrow in on the ones that are in the path of the ray.

Once we have a successful hit against one of the individual columns, we get the actual terrain height values at each corner of the column, shrink the column so it tightly encloses the four points, ray cast to test for a hit with the shrunken column, then finally triangulate and ray cast against the two triangles.

This is all done in an order that guarantees the first successful hit with a triangle will also be the nearest hit for the entire heightmap.

Reply:
Ohhhhh, thx for the explanation! Putting more of this description ("a grid of columns...") into the comments would be great, what you've written above makes it perfectly clear to me what's going on now.